man/fatcat.1
npage for fatcat

.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..

.\" Contact g.passault@gmail.com to correct errors or typos.
.TH man 1 "28 October 2017" "1.0.6" "fatcat man page"
.SH NAME
fatcat \- FAT filesystem explore, extract, repair, and forensic tool
.SH SYNOPSIS
fatcat disk.img [options]
.SH DESCRIPTION
fatcat is a standalone tool that allow you to explore, extract, repair and forensic FAT filesystems
It currently supports FAT12, FAT16 and FAT32.
.SH OPTIONS
You should provide an option to fatcat, or you will get the help menu

.PP
\fB\-i\fP
.RS 4
Display information about the FAT filesystem
.RE

.PP
\fB\-l path [\-d]\fP
.RS 4
Lists the entries of the directory \fBpath\fP
If \fB\-d\fP is present, deleted files will be listed.
.RE

.PP
\fB\-L cluster [\-d]\fP
.RS 4
Same as \fB\-l\fP, but using the \fBcluster\fP cluster number as source.
If \fB\-d\fP is present, deleted files will be listed.
.RE

.PP
\fB\-r path\fP
.RS 4
Reads the file given by \fbpath\fP
.RE

.PP
\fB\-R cluster [\-s size]\fP
.RS 4
Reads the file starting at the \fBcluster\fP cluster number. If \fBsize\fP is provied,
this will only read \fBsize\fP bytes.
.RE

.PP
\fB\-x target [\-c cluster] [\-d]\fP
.RS 4
Extract all the files to the \fBtarget\fP directory. If you provide a cluster with
\fB\-c\fP, the extract will start with the given cluster instead that the root directory.
If \fB\-d\fP is present, deleted files will be extracted.
.RE

.PP
\fB\-z, \-S\fP
.RS 4
\fB\-z\fP will write all your unallocated data to zero, and \fB\-S\fP will write all your
unallocated data to random bytes. this will for instance cause your deleted files to
be unreadable.
.RE

.PP
\fB\-@ cluster\fP
.RS 4
This will display information about the given \fBcluster\fP. It will display its address, which
is the offset of the cluster in the image, and the FAT entries (next cluster, unallocated
or end of cluster). A list of cluster chains is displayed if \fB-n\fB is used.
.RE

.PP
\fB\-n\fP
.RS 4
List cluster chains when displaying cluster information.
.RE

.PP
\fB\-2\fP
.RS 4
Compares the two file allocation tables and produce a full diff. This can be used to check that
the disk is not corrupted, and have a look to it before trying to merge it with \fB\-m\fP.
.RE

.PP
\fB\-m\fP
.RS 4
Merges the two file allocation tables. This will only keep the allocated entries from on or the
other table.
.RE

.PP
\fB\-b backupfile [\-t table]\fP
.RS 4
Backups your FAT tables to the \fBbackupfile\fP file. You can specify with \fB\-t\fP the 
table(s) you want to backup (0:both, 1:first, 2:second). You can then apply the FATs
using \fB\-p\fP.
.RE

.PP
\fB\-p backupfile [\-t table]\fP
.RS 4
Patch your FAT table using \fBbackupfile\fP previously backuped file (using \fB\-b\fP).
You can use \fB\-t\fP to specify the table(s) you want to patch (0: both, 1:first, 2:second).
.RE

.PP
\fB\-w cluster \-v value [\-t table]\fP
.RS 4
Writes the \fBcluster\fP entry in the FAT to \fBvalue\fP. You can specify the table using
\fB-t\fP (0:both, 1:first, 2: second).
.RE

.PP
\fB\-o\fP
.RS 4
Search for orphaned files on the disk. This will produce a log listing files, directories
and entries that are found. See \fB\-L\fP, \fB\-R\fP and \fB\-x\fP to access those files and
directories.
.RE

.PP
\fB\-f\fP
.RS 4
Walks the directories from the root (/) and try to fix unallocated files and directories
FAT table.
.RE

.PP
\fB\-e path [\-c cluster] [\-s size]\fP
.RS 4
Display information about the entry of the \fBpath\fP file or directory. You can edit its
cluster or size reference using \fB\-c\fP and \fB\-s\fP.
.RE

.PP
\fB\-k cluster\fP
.RS 4
Walks the directories from the root (/) and search an entry referencing the given \fBcluster\fP.
.RE

.SH EXAMPLES
You can explore your disk using \fB\-l\fP:

$ fatcat disk.img \-l /

And enter directories:

$ fatcat disk.img \-l /some/dir/

You can read a file using \fB\-R\fP:

.Vb 5
\&$ fatcat disk.img \-r /hello.txt 
\&Hello world!
\&$ fatcat disk.img \-r /picture.png > out.png
.Ve

You can also read files, including deleted ones:

$ fatcat disk.img \-l / \-d

And extract all the files to a target directory:

.Vb 5
\&$ mkdir output/
\&$ fatcat disk.img \-x output/
.Ve

Let's have a look at the listing:

.Vb 5
\&$ fatcat hello-world.img \-l /
\&Listing path /
\&Directory cluster: 2
\&f 25/10/2013 13:30:06  hello.txt                      c=3 s=13 (13B)
\&d 25/10/2013 13:30:46  files/                         c=4
.Ve

The cluster of the \fBfiles\fP directory is \fB4\fP, this means that we
can list it with \fB\-L 4\fP:

.Vb 5
\&$ fatcat hello-world.img \-L 4
\&Listing cluster 4
\&Directory cluster: 4
\&d 25/10/2013 13:30:22  ./                             c=4
\&d 25/10/2013 13:30:22  ../                            c=0
\&f 25/10/2013 13:30:46  other_file.txt                 c=5 s=29 (29B)
.Ve

The cluster of the \fBother_file.txt\fP file is 5, and its size is 29bytes,
we can then read it using \fB\-R 5 \-s 29\fP:

.Vb 5
\&$ fatcat hello-world.img \-R 5 \-s 29
\&Hello!
\&This is another file!
.Ve

For more examples and tutorials, have a look at the \fBfatcat\fP tutorial and examples
at: <\fBhttps://github.com/Gregwar/fatcat/blob/master/docs/index.md\fR>

.SH SEE ALSO
fatattr(1), mkdosfs(8)
.SH BUGS
No known bugs.
.SH AUTHOR
Grégoire Passault (g.passault@gmail.com)


debian/changelog
fatcat (1.1.1-1) UNRELEASED; urgency=medium

  * Output 8DOT3 short filenames whenever possible
  * Fix FAT12 determination
  * Writing error to stderr and exiting with error code
  * Added support for json output when listing entries
  * Fixing clusterAddress using temporary 32bit sized int (see debian bug
    879930)

 -- Grégoire Passault <g.passault@gmail.com>  Sat, 18 Jan 2022 09:20:00 +0200

fatcat (1.0.7-1) unstable; urgency=medium

  * Fixing line breaks in man page
  * Fixing possible cluster truncating (see #879930)

 -- Grégoire Passault <g.passault@gmail.com>  Sat, 28 Oct 2017 11:24:47 +0200

fatcat (1.0.5-1) unstable; urgency=low

    * Initial release. (Closes: #728125).

 -- Grégoire Passault <g.passault@gmail.com>  Thu, 07 Aug 2014 20:09:11 +0100


src/fatcat.cpp
#include <stdlib.h>
#include<string.h>
#ifdef __APPLE__
#include <unistd.h>
#else
#ifdef __WIN__
#include <ctype.h>
#include "xgetopt/xgetopt.h"
#else
#include <argp.h>
#endif
#endif

#include <stdio.h>
#include <string>
#include <iostream>

#include <FatUtils.h>
#include <core/FatSystem.h>
#include <core/FatPath.h>
#include <table/FatBackup.h>
#include <table/FatDiff.h>
#include <analysis/FatChains.h>
#include <analysis/FatExtract.h>
#include <analysis/FatFix.h>
#include <analysis/FatSearch.h>

#define ATOU(i) ((unsigned int)atoi(i))

using namespace std;

void usage()
{
    cout << "fatcat v1.1.0, Gregwar <g.passault@gmail.com>" << endl;
    cout << endl;
    cout << "Usage: fatcat disk.img [options]" << endl;
    cout << "  -i: display information about disk" << endl;
    cout << "  -O [offset]: global offset (may be partition place)" << endl;
    cout << "  -F [format]: output format (default, json)" << endl;
    cout << endl;
    cout << "Browsing & extracting:" << endl;
    cout << "  -l [dir]: list files and directories in the given path" << endl;
    cout << "  -L [cluster]: list files and directories in the given cluster" << endl;
    cout << "  -r [path]: reads the file given by the path" << endl;
    cout << "  -R [cluster]: reads the data from given cluster" << endl;
    cout << "  -s [size]: specify the size of data to read from the cluster" << endl;
    cout << "  -d: enable listing of deleted files" << endl;
    cout << "  -x [directory]: extract all files to a directory, deleted files included if -d" << endl;
    cout << "                  will start with rootDirectory, unless -c is provided" << endl;
    cout << "* -S: write scamble data in unallocated sectors" << endl;
    cout << "* -z: write scamble data in unallocated sectors" << endl;
    cout << endl;
    cout << "FAT Hacking" << endl;
    cout << "  -@ [cluster]: Get the cluster address and information" << endl;
    cout << "  -n: enable listing of cluster chains in cluster information" << endl;
    cout << "  -2: analysis & compare the 2 FATs" << endl;
    cout << "  -b [file]: backup the FATs (see -t)" << endl;
    cout << "* -p [file]: restore (patch) the FATs (see -t)" << endl;
    cout << "* -w [cluster] -v [value]: write next cluster (see -t)" << endl;
    cout << "  -t [table]: specify which table to write (0:both, 1:first, 2:second)" << endl;
    cout << "* -m: merge the FATs" << endl;
    cout << "  -o: search for orphan files and directories" << endl;
    cout << "* -f: try to fix reachable directories" << endl;
    cout << endl;
    cout << "Entries hacking" << endl;
    cout << "  -e [path]: sets the entry to hack, combined with:" << endl;
    cout << "* -c [cluster]: sets the entry cluster" << endl;
    cout << "* -s [size]: sets the entry size" << endl;
    cout << "* -a [attributes]: sets the entry attributes" << endl;
    cout << "  -k [cluster]: try to find an entry that point to that cluster" << endl;

    cout << endl;
    cout << "*: These flags writes on the disk, and may damage it, be careful" << endl;
    exit(EXIT_FAILURE);
}

int main(int argc, char *argv[])
{
    vector<string> arguments;
    char *image = NULL;
    int index;

    // -O offset
    unsigned long long globalOffset = 0;

    // -s, specify the size to be read
    unsigned int size = -1;

    // -i, display information about the disk
    bool infoFlag = false;

    // -l, list directories in the given path
    bool listFlag = false;
    string listPath;

    // -c, listing for a direct cluster
    bool listClusterFlag = false;
    unsigned int listCluster;

    // -r, reads a file
    bool readFlag = false;
    string readPath;

    // -R, reads from cluster file
    bool clusterRead = false;
    unsigned int cluster = 0;

    // -d, lists deleted
    bool listDeleted = false;

    // -x: extract
    bool extract = false;
    string extractDirectory;

    // -2: compare two fats
    bool compare = false;

    // -@: get the cluster address
    bool address = false;

    // -n: list cluster chains
    bool showChains = false;

    // -k: analysis the chains
    bool chains = false;

    // -b: backup the fats
    bool backup = false;
    bool patch = false;
    string backupFile;

    // -w: write next cluster
    bool writeNext = false;

    // -m: merge the FATs
    bool merge = false;

    // -v: value
    bool hasValue = false;
    unsigned int value;

    // -t: FAT table to write or read
    unsigned int table = 0;

    // -S: write random data in unallocated sectors
    bool scramble = false;
    bool zero = false;

    // -f: fix reachable
    bool fixReachable = false;

    // -e: entry hacking
    bool entry = false;
    string entryPath;
    bool clusterProvided = false;
    bool sizeProvided = false;
    int attributes = 0;
    bool attributesProvided = false;

    // -k: entry finder
    bool findEntry = false;
    OutputFormatType outputFormat = Default;

    // Parsing command line
    while ((index = getopt(argc, argv, "il:L:r:R:s:dc:hx:2@:nob:p:w:v:mt:Sze:O:fk:a:F:")) != -1) {
        switch (index) {
        case 'a':
            attributesProvided = true;
            attributes = atoi(optarg);
            break;
        case 'k':
            findEntry = true;
            cluster = atoi(optarg);
            break;
        case 'f':
            fixReachable = true;
            break;
        case 'O':
            globalOffset = atoll(optarg);
            break;
        case 'e':
            entry = true;
            entryPath = string(optarg);
            break;
        case 'z':
            zero = true;
            break;
        case 'S':
            scramble = true;
            break;
        case 't':
            table = ATOU(optarg);
            break;
        case 'm':
            merge = true;
            break;
        case 'v':
            hasValue = true;
            value = ATOU(optarg);
            break;
        case 'w':
            writeNext = true;
            cluster = ATOU(optarg);
            break;
        case '@':
            address = true;
            cluster = ATOU(optarg);
            break;
        case 'n':
            showChains = true;
            break;
        case 'o':
            chains = true;
            break;
        case 'i':
            infoFlag = true;
            break;
        case 'l':
            listFlag = true;
            listPath = string(optarg);
            break;
        case 'L':
            listClusterFlag = true;
            listCluster = ATOU(optarg);
            break;
        case 'r':
            readFlag = true;
            readPath = string(optarg);
            break;
        case 'R':
            clusterRead = true;
            cluster = ATOU(optarg);
            break;
        case 's':
            size = ATOU(optarg);
            sizeProvided = true;
            break;
        case 'd':
            listDeleted = true;
            break;
        case 'c':
            cluster = ATOU(optarg);
            clusterProvided = true;
            break;
        case 'x':
            extract = true;
            extractDirectory = string(optarg);
            break;
        case '2':
            compare = true;
            break;
        case 'b':
            backup = true;
            backupFile = string(optarg);
            break;
        case 'p':
            patch = true;
            backupFile = string(optarg);
            break;
        case 'h':
            usage();
            break;
        case 'F':
            if (strcmp(optarg, "json") == 0)
                outputFormat = Json;
            else if (strcmp(optarg, "default") == 0)
                outputFormat = Default;
            else
            {
                usage();
            }
            break;
        }
    }

    // Trying to get the FAT file or device
    if (optind != argc) {
        image = argv[optind];
    }

    if (!image) {
        usage();
    }

    // Getting extra arguments
    for (index=optind+1; index<argc; index++) {
        arguments.push_back(argv[index]);
    }

    // If the user did not required any actions
    if (!(infoFlag || listFlag || listClusterFlag || 
        readFlag || clusterRead || extract || compare || address ||
        chains || backup || patch || writeNext || merge ||
        scramble || zero || entry || fixReachable || findEntry)) {
        usage();
    }

    try {
        // Openning the image
        FatSystem fat(image, globalOffset, outputFormat);

        fat.setListDeleted(listDeleted);

        if (fat.init()) {
            if (infoFlag) {
                fat.infos();
            } else if (listFlag) {
                if (outputFormat != Json)
                    cout << "Listing path " << listPath << endl;
                FatPath path(listPath);
                fat.list(path);
            } else if (listClusterFlag) {
                if (outputFormat != Json)
                    cout << "Listing cluster " << listCluster << endl;
                fat.list(listCluster);
            } else if (readFlag) {
                FatPath path(readPath);
                fat.readFile(path);
            } else if (clusterRead) {
                fat.readFile(cluster, size);
            } else if (extract) {
                FatExtract extract(fat);
                extract.extract(cluster, extractDirectory, listDeleted);
            } else if (compare) {
                FatDiff diff(fat);
                diff.compare();
            } else if (address) {
                cout << "Cluster " << cluster << " address:" << endl;
                long long addr = fat.clusterAddress(cluster);
                printf("%llu (%016llx)\n", addr, addr);
                cout << "Next cluster:" << endl;
                int next1 = fat.nextCluster(cluster, 0);
                int next2 = fat.nextCluster(cluster, 1);
                printf("FAT1: %u (%08x)\n", next1, next1);
                printf("FAT2: %u (%08x)\n", next2, next2);

                FatChains chains(fat);
                list<Segment> segments;
                unsigned long long size = chains.chainSize(cluster, segments);
                printf("Chain size: %llu (%llu / %s)\n", size, size*fat.bytesPerCluster, prettySize(size*fat.bytesPerCluster).c_str());
                if (segments.size() < 2) {
                    printf("Chain is contiguous\n");
                } else {
                    printf("Chain is not contiguous\n");
                }

                if (showChains) {
                    list<Segment>::iterator it;
                    int num=1;
                    for (it=segments.begin(); it != segments.end(); num++,it++) {
                        int start = it->first;
                        int end = it->second;
                        int size = end - start + 1;
                        printf("Chain %d: %d..%d (%d clusters)\n", num, start, end, size);
                    }
                }
            } else if (chains) {
                FatChains chains(fat);
                chains.chainsAnalysis();
            } else if (fixReachable) {
                FatFix fix(fat);
                fix.fix();
            } else if (findEntry) {
                FatSearch search(fat);
                search.search(cluster);
            } else if (backup || patch) {
                FatBackup backupSystem(fat);
                
                if (backup) {
                    backupSystem.backup(backupFile, table);
                } else {
                    backupSystem.patch(backupFile, table);
                }
            } else if (writeNext) {
                if (!hasValue) {
                    throw string("You should provide a value with -v");
                }

                int prev = fat.nextCluster(cluster);
                printf("Writing next cluster of %u from %u to %u\n", cluster, prev, value);
                fat.enableWrite();

                if (table == 0 || table == 1) {
                    printf("Writing on FAT1\n");
                    fat.writeNextCluster(cluster, value, 0);
                } 
                if (table == 0 || table == 2) {
                    printf("Writing on FAT2\n");
                    fat.writeNextCluster(cluster, value, 1);
                } 
            } else if (merge) {
                FatDiff diff(fat);
                diff.merge();
            } else if (scramble) {
                fat.enableWrite();
                fat.rewriteUnallocated(1);
            } else if (zero) {
                fat.enableWrite();
                fat.rewriteUnallocated();
            } else if (entry) {
                cout << "Searching entry for " << entryPath << endl;
                FatPath path(entryPath);
                FatEntry entry;
                if (fat.findFile(path, entry)) {
                    printf("Entry address %016llx\n", entry.address);
                    printf("Attributes %02X\n", entry.attributes);
                    cout << "Found entry, cluster=" << entry.cluster;
                    if (entry.isDirectory()) {
                        cout << ", directory";
                    } else {
                        cout << ", file with size=" << entry.size << " (" << prettySize(entry.size) << ")";
                    }
                    cout << endl;

                    if (clusterProvided) {
                        cout << "Setting the cluster to " << cluster << endl;
                        entry.cluster = cluster;
                    }
                    if (sizeProvided) {
                        cout << "Setting the size to " << size << endl;
                        entry.size = size&0xffffffff;
                    }
                    if (attributesProvided) {
                        cout << "Setting attributes to " << attributes << endl;
                        entry.attributes = attributes;
                    }
                    if (clusterProvided || sizeProvided || attributesProvided) {
                        entry.updateData();
                        fat.enableWrite();
                        string data = entry.data;
                        fat.writeData(entry.address, data.c_str(), entry.data.size());
                    }
                } else {
                    cerr << "Entry not found." << endl;
                    exit(EXIT_FAILURE);
                }
            }
        } else {
            cerr << "! Failed to init the FAT filesystem" << endl;
            exit(EXIT_FAILURE);
        }
    } catch (string error) {
        cerr << "Error: " << error << endl;
        exit(EXIT_FAILURE);
    }

    exit(EXIT_SUCCESS);
}
